CMakeLists.txt

```diff
@@ -594,7 +594,16 @@ if(SWIFT_HOST_VARIANT_SDK)
   set(SWIFT_HOST_VARIANT_SDK_default "${SWIFT_HOST_VARIANT_SDK}")
 else()
   if("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
-    set(SWIFT_HOST_VARIANT_SDK_default "LINUX")
+    execute_process(
```

Comment:
Please leave a comment explaining that this won't break the Windows -> Linux cross-compilation.

Reply:
Are we sure that it won't? I had not considered that cross-compilation combo, but I guess it won't work if this code path is hit, ie does Windows have uname?

If you try to cross-compile by setting the SWIFT_HOST_VARIANT, then this isn't hit, no problem. However, if you try to cross-compile by setting CMAKE_SYSTEM_NAME to "Linux" on Windows instead, I don't think this will work.

I'm not sure any of this matters since there doesn't appear to be any support for cross-compiling the host tools from Windows to linux right now.

Reply:
Windows does not have `uname` and we are building the android components on Windows too now :-).

Reply:
>Windows does not have uname and we are building the android components on Windows too now :-).

Not the host tools like the Swift compiler though, so that situation's different.

In fact, given the inconsistent usage of CMAKE_SYSTEM_NAME and SWIFT_HOST_VARIANT in this repo, I bet nobody has ever actually tried cross-compiling the Swift host tools to _any_ OS.

Reply:
@buttaface - if you ignore the fact that I'm cross-compiling from Linux to Windows, and Android and Windows to Android, and macOS to Linux and Windows :-).

Reply:
@drodriguez,
>I think you just lost a bet

Got an example? Remember that we're talking here about cross-compiling the Swift compiler and other host tools for a different OS, eg cross-compiling the bin/swift compiler itself on Windows for linux, then running that Windows-built bin/swift ELF binary on linux.

>But I think that changing this to check for CMAKE_HOST_SYSTEM_NAME, which should be Linux even when cross-compiling from Windows into Linux or any other combination.

Yes and no. That host variable won't be set to Linux when cross-compiling from Windows. However, I do think it would make more sense if the variable was CMAKE_HOST_SYSTEM_NAME here.

@compnerd,
>if you ignore the fact that I'm cross-compiling from Linux to Windows, and Android and Windows to Android, and macOS to Linux and Windows

Again, are you cross-compiling the host tools like the bin/swift compiler or simply cross-compiling the stdlib or other Swift files? While the former could be done, given the inconsistent support for it in the current CMake config, I suspect you're only doing the latter.

Reply:
I am cross-compiling the host tools *AND* the standard library.  This is pretty well documented even.

Reply:
>I am cross-compiling the host tools AND the standard library. This is pretty well documented even.

Link me where it's documented? I'd like to know what variables were set to configure those cross-builds, given the inconsistencies I've pointed out, just as I asked @drodriguez in another thread.

Reply:
https://github.com/apple/swift/blob/master/docs/WindowsCrossCompile.md is that, I have the cross-compiling in a separate repo (toolchain-instfrastructure on github, feel free to check that out)

Reply:
Thanks, I just took a look at both, but they seem to be about cross-compiling the Swift runtime and standard library for Windows, not cross-compiling the Swift compiler itself and other host tools. If I missed it, can you point to where in your toolchain-infrastructure repo you're configuring cross-compiling the Swift host tools?

Reply:
They are for both the host tools **AND** the runtime, not just the runtime.  The patches for supporting just the cross-compile of the runtime aren't posted yet.

Reply:
I don't need to read the whole thing: I just spent a couple minutes looking at the top-level Makefile and running github searches for the relevant variables, [CMAKE_SYSTEM_NAME](https://github.com/compnerd/toolchain-infrastructure/search?q=CMAKE_SYSTEM_NAME&unscoped_q=CMAKE_SYSTEM_NAME) and [SWIFT_HOST_VARIANT_SDK](https://github.com/compnerd/toolchain-infrastructure/search?q=SWIFT_HOST_VARIANT_SDK). It appears that [you simply set CMAKE_SYSTEM_NAME](https://github.com/compnerd/toolchain-infrastructure/blob/master/Makefile#L63) to [the host OS](https://github.com/compnerd/toolchain-infrastructure/blob/master/Makefile#L6), in which case you're __not__ cross-compiling the host tools, just like I said.

I suppose it's possible you're overriding the Makefile variables when you run it on the command-line, in which case it's pointless to tell me to look at any of those files, as what matters is what command-line variables you're passing in, not what's in those files.

It looks to me that you don't really understand these matters, with how you kept talking about the NDK when I kept saying I wasn't using it and now point to docs and scripts that don't reference building the host tools at all, all the while claiming they do.

Reply:
I think you may want to actually look at my contributions before assuming that it is *I* who doesn't understand this matter ;-).

Reply:
Let's don't take that road. This can not end well for anyone. I will leave a longer answer in the main thread.

@buttaface: please believe us, we have cross-compiled the compiler to other platforms.

Reply:
@compnerd, there's no BuildHost variable used in that repo. :smile: Another sign that you're not cross-compiling the Swift compiler itself is that you'd have to also cross-compile LLVM itself, ie have a Windows version of the LLVM libraries built or downloaded on linux to link the Swift build for Windows against, but there's no mention of using a different LLVM in that repo.

Perhaps you have done what you claim, ie cross-compiling the Swift compiler itself for Windows on a linux host, but the links you gave don't support that claim. Maybe you have additional patches or CMake config locally that you're using for cross-compiling the Swift compiler itself and haven't pushed them to github.

Anyway, getting back to the original point of this thread, the Swift CMake config for building the host tools currently uses two variables, CMAKE_SYSTEM_NAME and SWIFT_HOST_VARIANT, to signify pretty much the same thing. That will work as long as they're both set to the same OS, but is error-prone because someone might accidentally set them to different OS's and this Swift host build will then go haywire.